### ISSUE TITLE
Finish migrating to GitHub

### ISSUE THREAD

**Issue by blackwind** [EventGhost/EventGhost#2]
Unfortunately, the user who was migrating our SVN repository to GitHub bailed before completion, and no one else currently on the team is familiar enough with Git to pick up where he left off. If you think you have the know-how to get us over the finish line, please post a comment below. Until the migration is complete, no pull requests will be accepted at this repository.

**Comment by per1234** [comment]
There is some infomation posted [here](http://eventghost.net/forum/viewtopic.php?t=7382#p38800) by Pako that is relevant to this issue
>I am not the author or owner EventGhost. So I can not be against the move to Github, if it is the desire of most developers. However, I have no pleasure in it. SVN repository suits me because it is very simple. I do not know how to work with Github.
If I am not mistaken, (currently) I'm the only person who is able to create a new EventGhost build. I have no pleasure in this fact.
But it has one consequence.
I have only one computer, which is set for this task. It should also be noted that the compilation EventGhost is based on the fact that the repository is type of SVN and after moving to Github, it is necessary to adjust the appropriate code.
So I have this attitude:
1) You can work on moving to Github.
2) Once you are able to release new builds from Github, then I can leave the SVN repository and add to you. Until then, I can not interfere with the computer settings, even in the code that relates to the compilation from the SVN repository.

Also for information on the original attempt at the migration: http://eventghost.net/forum/viewtopic.php?t=5385

I have no experience with migrating from SVN to Git but would really like to see this happen(assuming the developers will actually make the switch to using this repo once the migration is complete). If there is any sort of grunt work needed I'm happy to volunteer to do it, just let me know.

**Comment by Christoph-Hennemann** [comment]
Unfortunately I do not have any experience in migrating SVN to Git either. But I found a nice tutorial that might help you a lot.

It can be found [here](https://www.atlassian.com/git/tutorials/migrating-overview/).

It specifically relates to Bitbucket (another git client like Github) but that should be no different to migrating to Github itself. It is suggested to select one projectmember as the migration lead to be the only one commiting to the newly created git repository. All other team members treat the git repository as read only during the migration phase. With this approch everyone has the time to become familiar with the git flow until the migration happens at once.

I hope this helps.

**Comment by CarsonF** [comment]
I deeply regret leaving this unfinished. My original goal was to clean up the svn commits and merge some branches into master. Unfortunately, this got out of hand quickly and created the merge conflicts were vast enough to force me to dive into the codebase. Hindsight is 2020. Honestly, I would just pull in `pako2`'s branch, if that's still the latest one updated, and make that one the master. However, I don't think this can be done with GitHub's PRs.

**Comment by CarsonF** [comment]
Since I have write access I'll see if I can do a straight import from SVN to another branch. If all looks good we can make that one master.

**Comment by blackwind** [comment]
Thanks for checking in, Carson. No hard feelings or anything -- we all start things we don't get around to finishing. Hope you're doing well. :)

GitHub has its own [importer](https://import.github.com/new) now, and I went ahead and ran it against SVN earlier today. @topic2k is on the case. It's my hope that we can still complete the migration in full as originally intended.

**Comment by CarsonF** [comment]
@blackwind Thank you, I'm doing well. Are we wanting to use the _importer_? It looked like from #3 that there were only 2 commits which isn't right. I'm running through the link @chennemann posted. I'm hesitient about the _importer_ working the way we want it to because really we want the root to be `branches/pako2/plugin-install`. I'm worried the _importer_ will leave us with `plugin-install` and `not-included-plugins` in the root.

**Comment by CarsonF** [comment]
Maybe one commit to move everything from `plugin-install` to root instead of rewriting history would be better though

**Comment by CarsonF** [comment]
Ah see, I'm remembering now. There's some derpy stuff I was trying to avoid. I pushed pako's branch to https://github.com/EventGhost/EventGhost/tree/pako2. There are only 224 commits, instead of the 1400 in master, because of this 3587b2f

**Comment by blackwind** [comment]
#3 was solely his work and not the importer. You can take a look at the importer's initial results in [EventGhost-import](https://github.com/EventGhost/EventGhost-import). The branches are, indeed, separated into those folders, but master isn't. Logically, I would expect a merge into master to slot the files into the locations we want, but then, source control really isn't my forte. Worst case scenario, though, an extra commit should do the trick if needed, as you note.

**Comment by blackwind** [comment]
It sure would be nice if GitHub could automatically resolve conflicts. This whole thing is making my head hurt.

**Comment by CarsonF** [comment]
I think the bigger problem, one of a couple reasons this got dropped by me, is that no devs knew git. So even if I did get it finished, it would be hard to migrate people to using it.

**Comment by blackwind** [comment]
I don't see that as being a huge deal with [GitHub Desktop](https://desktop.github.com/), TortoiseGit, Visual Studio Code, and so forth available. I know I've even contributed to a number of projects by simply copy/pasting code into GitHub's text editor here on the website. Once we get over the initial hump of merging these massive, multi-year branches, there'll be no further need for any complicated command-line wizardry.

**Comment by CarsonF** [comment]
Fair point. I'm going to take what you have from the import repo and see if I can spend a couple hours rebasing the `plugin-install/` out. What should be done with the `not-included-plugins`? In an ideal world, most of the plugins should be in their own repo. Maybe for now they can just live a separate branch?

**Comment by blackwind** [comment]
Remove not-included-plugins completely. Since Pako has chosen not to make these core EventGhost plugins, they're effectively Pako plugins and should each have a separate repo under his account. When I speak to him about the transition, I'll recommend that he uses the importer himself to accomplish that.

**Comment by CarsonF** [comment]
Will do. Are you on IRC or Slack or something I can ping you about these quick questions?

**Comment by blackwind** [comment]
irc://irc.efnet.org/blackwind

**Comment by CarsonF** [comment]
Ok I've pushed the [master-updated](https://github.com/EventGhost/EventGhost/tree/master-updated) branch. It was based on the [pako2 branch from EventGhost-import](https://github.com/EventGhost/EventGhost/tree/pako2). I started rebasing [here](https://github.com/EventGhost/EventGhost/commit/da24bd2288999ed5d1d30dee9b4de6f99a335f4c) where the `plugin-install` and `not-included-plugins` folders started. I basically just moved everything from `plugin-install` to _root_ and deleted the commits for `not-included-plugins`. 

History is mostly intact, which was my main concern.

**Comment by CarsonF** [comment]
Pushed that branch to `master` and deleted the used ones. Left `bitmonster` branch for reference. Code could be cherry picked from it, but it shouldn't be merged.

@blackwind close?

**Comment by blackwind** [comment]
Waiting on @topic2k's build script pull request before closing. We're not officially migrated until Pako's requirements are met.

**Comment by CarsonF** [comment]
Ah right the build script! Also for versioning, it would be a good time to switch to semantic versioning.
maybe `v5.0.0`?

**Comment by blackwind** [comment]
Agreed. I believe v0.5.0 is what we agreed on at the forum years back. I'll be digging in and making that change and many others once I can successfully build.

**Comment by per1234** [comment]
I don't know what would be involved in implementing something similar for this project but I work with the [Arduino repository](https://github.com/arduino/Arduino) and they do an automatic hourly build if there have been new commits to the repository and there is a link to the build on the download page of the website. Also, the ArduinoBot will build pull requests when a collaborator requests it in a comment. This makes it easy for beta testers because they don't have to do the builds themselves and makes sure everyone is working with the same files.

**Comment by CarsonF** [comment]
Cool. As I said on IRC, I've stepped away from this project for lack of time and interest. But feel free to ping me if you need help with any git/github stuff. Happy to help out in that capacity.

One thing to think about is writing a PR how-to guide.
http://www.thinkful.com/learn/github-pull-request-tutorial

**Comment by blackwind** [comment]
@per1234: I was investigating exactly that last night. It's called "continuous integration", and I'm leaning towards [Travis CI](https://travis-ci.org/) as our provider unless someone with more experience cares to weigh in.

@CarsonF: Cheers! And thanks for everything!

**Comment by per1234** [comment]
>One thing to think about is writing a PR how-to guide.

If you put this in a file named CONTRIBUTING.md or CONTRIBUTING then there will be a link to that file displayed whenever someone starts to submit an issue or pull request: https://help.github.com/articles/setting-guidelines-for-repository-contributors/ This could also contain guidelines for submitting issue reports such as:
- Search the existing issues and pull requests to avoid duplicates.
- Test the hourly build to see if the issue has already been fixed
- State the version of Windows you are using
- The issue tracker is for bug reports only, use the EventGhost forum for support requests

This could help improve the quality of contributions while reducing the support load on the developers.

**Comment by aderusha** [comment]
I don't have much to add here except THANK YOU EVERYBODY.  This move has been a long time coming, and I know it isn't easy.  As a regular user, I really appreciate everyone's work here!

**Comment by topic2k** [comment]
CarsonF wrote:
>  It looked like from #3 that there were only 2 commits which isn't right. 

blackwind wrote:
> #3 was solely his work and not the importer.

Yes, right. I made the changes manually to the master and didn't used the importer.

---

The repo has now changed. If i understand right, the master branch is now what was pako2/plugin-install? Then i will do the build tool based on the new master.

Before i start with integrating the build tool, i would ask, if it would be ok to split the repo into multiple repos. That would it make simpler to make the build tool. Something like:
EventGhost organization:
  - EventGhost
  - build tool
  - extension: cFunctions
  - extension: dxJoystick
  - extension: WinUsbWrapper
  - Plugins (all)     or
  - plugin_A
  - plugin_B
  - plugin_C  etc.
Have a look at https://github.com/humpibumpi which i used for testing. 

If thats ok, i would first start with just the build system for the installer, which will use the lokal repo on the devs machine to build the installer. After that i will (re)add the task for updating the CHANGELOG.txt based on commit messages. These i will make editable before creating the installer.
I didn't yet looked at the tasks for creating the website, html docs, upload and sync website tasks. These i will leave for later.

**Comment by CarsonF** [comment]
I agree that a lot of plugins could be in their own separate repos.

However, I don't the build tool, and maybe extensions, should be separated. It makes it harder to develop locally.

**Comment by topic2k** [comment]
Sorry, i don't fully understand: do you mean, ahving separates repos would make it harder to develop locally?

**Comment by CarsonF** [comment]
Yeah. Build tools are never separated from source code repos.

Dependencies are separated. As long as there is a tool to pull in these dependencies (package manager, aka `pip`). Not sure how this would work with the C extensions, that's why I said leave them here for now.

**Comment by topic2k** [comment]
Thank you for clarification.

**Comment by CarsonF** [comment]
Sure. As I starting point, I would just get the build tool working. Dependencies can be separated later. It's important to get the migration finished ASAP so the GIT version doesn't deviate from the SVN version...again.

**Comment by topic2k** [comment]
ok, can we have a branch 'develop' where i can make a PR? As i learned there should be a develop branch where the changes will be tested before they will be merged into master.

**Comment by blackwind** [comment]
Long-term, almost all non-essential plugins will be split out of the distribution when the [online plugin repository](https://github.com/EventGhost/EventGhost/issues/4) hits. Long-term, there's a huge mess of C/C++ source code all over the distribution that should be, if you'll pardon the pun, brought up to code (project files the same version of Visual Studio, same build environments, etc.) and consolidated in one location, which might be the extensions folder, a separate repo, or multiple separate repos. Short-term, though, I can't echo this loudly enough:

> It's important to get the migration finished ASAP so the GIT version doesn't deviate from the SVN version...again.

Regarding the "develop" branch, my preference would be to simply test using your fork until your work is ready to be accepted as a pull request, but I'll defer to Carson's expertise on best Git practices if his opinion differs.

**Comment by CarsonF** [comment]
Let me explain where `develop` branches are useful. I work as a web developer, and as such our product/websites don't have _versions_. They aren't deployed to clients so we can update them whenever we want. We do not have any tags or branches for version numbers. We have two (main) branches `master` and `develop`. `master` is always represents our production environment, and `develop` represents our staging environment (where QA / other internal people can see/approve changes). Actual code being developed happens on `feature/*` or `bugfix/*` branches. This is typically how it is for deployed systems.

However, for libraries and products where clients can request a specific version this doesn't make sense. This is typically because the product needs to support multiple versions at once. For these, `master` is basically viewed as `develop`. It is unstable and unreleased. Once code is stabilized, it is pushed to a _release_ branch such as `release/5.0` or `5.0`. `master` then becomes the next unstable release. In the above example, `master` would alias what would become `5.1`. Tags or releases are then created off of the release branch. In this example, `5.0.0` would be tagged.

Say you've just added a feature to `master` (5.1). It is mostly stable but needs some bug testing. Then someone else finds a bug in `5.0.0` release. A bugfix would be PR'ed and merged to the `release/5.0` branch. Then `release/5.0` would be merged into `master` branch, so the bug is fixed there as well. Then the `HEAD` (latest commit) of the `release/5.0` would be tagged as `5.0.1`.

----

Could a `develop` branch still work with this workflow? Yes it could. [_A successful git branching model_](http://nvie.com/posts/a-successful-git-branching-model/) includes it. If you look at the image in that article, I've deviated by removing the `master` branch (each commit to that branch is tagged making it kinda pointless) and renaming `develop` branch to `master`.

**Comment by blackwind** [comment]
Thanks for the write-up, Carson. I've been shouting from the rooftops as loudly and frequently as I can that I know absolutely nothing of source control, but everything you just said lines up with the limited understanding I've gleaned from observing other projects. When we begin work on far-off, still-incomplete projects like the plugin repo that we don't want going out half-baked in minor releases, that's where I think we could benefit from an alternate branch. I don't view the build script as such a project. When it's done, we'll accept the pull request in master; until then, it can be tested via the fork.

**Comment by topic2k** [comment]
I'm nearly done. The installer is build from repository from github. Currently it uses my branch 'develop' to build because of the new version numbering. You can change it in `tools/builder/__init__.py`, line 62 and 63.

As a new dependency [pygit2](http://www.pygit2.org/) is needed.

Next i would like to have an automated commit of the changelog, creating a tag/release and uploading of the installer. But currently i'm stuck at doing commit/push from within the script.

While building the chm docs i get some warnings, but the file is build. Haven't looked further at it.

**Comment by blackwind** [comment]
Incredible work on both this and the Python 2.7 upgrade, topix. You're single-handedly pulling EventGhost into the 21st century, and we can't thank you enough.

For the automation stuff, be sure to make it possible to spin up a minimal build *without* the automation, and be sure to make minimal builds the default. We obviously don't want Travis CI releasing new builds every time it runs, for example, nor do we want anyone releasing a build accidentally when they only wanted to test.

For the docs, md5sum your result against what's currently included in the installer. If they match, good enough. I wouldn't be surprised if it's been generating warnings all along.

For the pull request, send to master. If you're confident that your python27 branch is ready to go, send us that one instead. I'll take a look at recompiling those .dlls and .pyds against Stackless 2.7.9 later today, as I highly doubt the hex edit will be sufficient.

**Comment by topic2k** [comment]
No, i'm not working on `setup.py`, just the build tool. Where do you have these from?

About the docs: do you have a app to do such md5sum checks?

**Comment by blackwind** [comment]
I saw you modified `extensions\_dxJoystick\setup.py` to change "dxerr9" to "dxerr", which is required to successfully build _dxJoystick, so I assumed you were on the case. I'll take another look tonight after I wade through and merge the pull request.

For checksums, [Cygwin](https://cygwin.com/install.html) contains the full suite of GNU checksum tools (`md5sum`, `sha1sum`, et al). Alternately, if you prefer an integrated Windows solution, I recommend [HashTab](http://implbits.com/products/hashtab/).

**Comment by topic2k** [comment]
Oh, I thought i left those changes out. I tried building it VS2015 and with the new SDK `dxer9` is renamed to `dxerr`.

I will look at those tools, but meanwhile i spent a little time on the docs generation and made some progress.
